### PR TITLE
C++ API TransformerEncoderLayer

### PR DESCRIPTION
**Pull request by glaringlee**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42633 C++ API TransformerEncoderLayer**

Differential Revision: [D22994332](https://our.internmc.facebook.com/intern/diff/D22994332)

### CODE REVIEW

**Comment by glaringlee** [comment]
@zhangguanheng66 Can you take a look at the logic in this C++ impl? thx
@yf225 Can you take a look at the c++ impl or point me to someone can look into this? thx

File: test/cpp/api/transformer.cpp [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  model.get()->options.activation(torch::kGELU);
+  encoder_input = torch::tensor({{{20, 30, 40, 50}}}, tensor_options);
+  result = model(encoder_input).detach();
+  ref_output = torch::tensor({{{2.249815, 0.131006, -0.702199, 0.177868}}}, tensor_options);
+  ASSERT_EQ(result.sizes(), ref_output.sizes());
+  ASSERT_TRUE(torch::allclose(result, ref_output, 1e-7, 1e-5, /*equal_nan=*/true));
+
+  // gelu test case 2
+  encoder_input = torch::tensor({
+    {{0.7462, 0.6653, 0.5679, 0.4891}, {0.5387, 0.1655, 0.3565, 0.0471}},
```
Comment (zhangguanheng66) [inline test/cpp/api/transformer.cpp:127]: A quick question. Are those deterministic tests consistent with those in python version?

Comment (glaringlee) [reply test/cpp/api/transformer.cpp:127]: yes, I copied them from python test directly, same number, same precision.

**Review by yf225 — APPROVED** [review]

**Comment by EddieMataEwy** [comment]
TransformerEncoderLayer options do not match current Pytorch options, particularly batch_first and norm_first.
Is there any plan to add the options to the C++ API?
Thank you!

**Pull request closed without merging**